extensions-core/druid-lookups/src/main/java/io/druid/server/lookup/LoadingLookupFactory.java

```diff
@@ -0,0 +1,144 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.lookup;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.metamx.common.logger.Logger;
+import io.druid.query.lookup.LookupExtractorFactory;
+import io.druid.query.lookup.LookupIntrospectHandler;
+import io.druid.server.lookup.cache.loading.LoadingCache;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@JsonTypeName("loadingLookup")
+public class LoadingLookupFactory implements LookupExtractorFactory
+{
+  private final static Logger LOGGER = new Logger(LoadingLookupFactory.class);
+
+  @JsonProperty("dataFetcher")
+  private final DataFetcher<String, String> dataFetcher;
+
+  @JsonProperty("loadingCacheSpec")
+  private final LoadingCache<String, String> loadingCache;
+
+  @JsonProperty("reverseLoadingCacheSpec")
+  private final LoadingCache<String, List<String>> reverseLoadingCache;
+
+  private final LoadingLookup loadingLookup;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  public LoadingLookupFactory(
+      @JsonProperty("dataFetcher") DataFetcher dataFetcher,
+      @JsonProperty("loadingCacheSpec") LoadingCache<String, String> loadingCache,
+      @JsonProperty("reverseLoadingCacheSpec") LoadingCache<String, List<String>> reverseLoadingCache
+  )
+  {
+    this.dataFetcher = Preconditions.checkNotNull(dataFetcher);
+    this.loadingCache = Preconditions.checkNotNull(loadingCache);
+    this.reverseLoadingCache = Preconditions.checkNotNull(reverseLoadingCache);
+    this.loadingLookup = new LoadingLookup(this.dataFetcher, this.loadingCache, this.reverseLoadingCache);
+  }
+
+  @VisibleForTesting
+  protected LoadingLookupFactory(
+      DataFetcher dataFetcher,
+      LoadingCache<String, String> loadingCache,
+      LoadingCache<String, List<String>> reverseLoadingCache,
+      LoadingLookup loadingLookup
+  )
+  {
+    this.dataFetcher = dataFetcher;
+    this.loadingCache = loadingCache;
+    this.reverseLoadingCache = reverseLoadingCache;
+    this.loadingLookup = loadingLookup;
+  }
+
+  @Override
+  public boolean start()
+  {
+    if (!started.get()) {
+      started.set(loadingLookup.isOpen());
+      LOGGER.info("starting loading lookup");
+    }
+    return started.get();
+  }
+
+  @Override
+  public boolean close()
+  {
+    if (started.getAndSet(false)) {
+      LOGGER.info("closing loading lookup");
+      loadingLookup.close();
+    }
+    return !started.get();
+  }
+
+  @Override
+  public boolean replaces(
+      @Nullable LookupExtractorFactory lookupExtractorFactory
+  )
+  {
+    if (lookupExtractorFactory == null) return false;
```

Comment:
the method says "does this object replace the other", if `this` is null, the method should never be called. If `other` is null, it should be replaced by `this` most likely